### PR TITLE
Replication issues 515, 518, 522, 535 - Fixes, State Logic restructure

### PR DESCRIPTION
**Pull request by ChristianAtDell**
# Description
A string of defects have been addressed, and with them, some significant refactoring has been done on replication.go code. Changes are described below, in order of significance. 

1. Planned Failover operation has been reworked to follow documented PowerScale failover steps and no longer call allow_write_revert. This call was blocking writes on source side after failover, but was also deleting source side entirely (causing CSM issues #518 and potentially #515). Now, both source and target sides are able to write after a failover, and user must follow PowerScale recommended steps to stop writes on the source (which will delete source data). Failback operations are planned for future work to allow retaining writes on the source (FAILBACK_DISCARD_CHANGES) OR retaining data from the target (FAILBACK_LOCAL or FAILBACK_REMOTE) at the discretion of the user. 
2. Unplanned failover steps have been reworked to remove break_association calls and instead simply allow writes on target side after discussions with SyncIQ SME. Break association was an improper action here and would have prevented future failback work. It also caused unplanned failover to not work. (CSM issue #522)
3. Reprotect operation has been removed temporarily while it is under investigation. It calls allow_write_revert similar to previous Planned Failover operation and needs to be looked at. 
4. All state logic has been adjusted. Previous state logic would not update state to UNKNOWN in error-cases, improperly leaving state as SYNCHRONIZED. New state logic should be much more robust and handle error states gracefully. It is also documented much more thoroughly. 
5. 'Suspend' operation and state have been reworked. Previous Suspend logic involved creating a folder on the source storage array and check the status of the SyncIQ job to determine if it is currently enabled. New logic simply checks the SyncIQ job to see if it's enabled.
6. IsSource logic has been adjusted. Previous logic would set source to false if replication was suspended, which meant that Resume operation would never succeed, as it required one side to be labeled as source. This change addresses CSM issue #535. 
7. Removed NFS Export deletions from the replication-direction check logic. These deletions seemed to be targeted at refusing access to target PV while in synchronized replication state, but target PV should still be accessible in a read-only state. 
8. Some unit test cases have been removed in accordance with new state checking logic, because old unit tests did not account for failed states appropriately. They are under investigation and will be re-added when they have been reworked. 
9. Some unit tests have had their checking conditions adjusted to fit the error messages of new state checking logic. 
10. New unit test added to check unsupported actions. 

Major thanks to @santhoshatdell for all his work on these code changes alongside me. 

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| [CSM Replication # 535](https://github.com/dell/csm/issues/535) |
| [CSM Replication # 522](https://github.com/dell/csm/issues/522) |
| [CSM Replication # 518](https://github.com/dell/csm/issues/518) |
| [CSM Replication # 515](https://github.com/dell/csm/issues/515) |

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [X] I have not allowed coverage numbers to degenerate (80.2% on current main -> 80.2% on this PR)
- [ ] I have maintained at least 90% code coverage
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

Attached are two test logs. The first test verifies all standard and supported replication actions-- planned failover, unplanned failover, suspend, resume, sync. 

[PowerScale replication RG status repctl output.txt](https://github.com/dell/csi-powerscale/files/9992757/PowerScale.replication.RG.status.repctl.output.txt)

The second test verifies corner cases by blocking access to storage arrays to simulate outages. 

[PowerScale replication RG status repctl output - corner cases.txt](https://github.com/dell/csi-powerscale/files/9992758/PowerScale.replication.RG.status.repctl.output.-.corner.cases.txt)

Both test cases produced expected states on the associated Replication Group objects.

### CODE REVIEW

File: service/replication.go [hunk 1]
```diff
@@ -113,8 +111,8 @@ func (s *service) CreateRemoteVolume(ctx context.Context,
 		var quotaID string
 		quota, err := remoteIsiConfig.isiSvc.client.GetQuotaWithPath(ctx, exportPath)
 		log.Info("Get quota", quota)
-		if quota == nil {
-			log.Info("Remote quota doesn't exists, create it")
+		if err != nil {
```
Comment (donatwork) [inline service/replication.go:114]: It is possible that you may get an error for other reasons than if the quota does not exist. For example the REST call may fail for a number of reasons. May be worth refactoring this in the future to handle such cases. If the quota did exist but the REST call failed then the subsequent operation to create the quota will also fail. How does the API behave if there is a call to get a quota that does not exist, is there a 404 or similar error?

File: service/replication.go [hunk 2]
```diff
@@ -475,131 +469,92 @@ func (s *service) GetStorageProtectionGroupStatus(ctx context.Context, req *csie
...
+		return resp, status.Errorf(codes.Internal, errMsg)
 	}
 
-	log.Info("trying to get replication direction")
+	log.Info("Trying to get replication direction")
 	source := false
-	if s1P.Enabled {
+	if linkState != csiext.StorageProtectionGroupStatus_FAILEDOVER && // no side can be source when in failed over state
+		(localP.Enabled || // when synchronized
+			(!remoteP.Enabled && localTP.FailoverFailbackState == "writes_enabled" && remoteTP.FailoverFailbackState == "writes_disabled")) { // when suspended (source side)
```
Comment (donatwork) [inline service/replication.go:545]: Minor cleanup issue, these two strings are used in a lot of places. May be worth consolidating into a set of constants.

**Review by donatwork — APPROVED** [review]
Need someone on the CSI PowerScale team to also review.

**Review by harshaatdell — APPROVED** [review]
Aweosme job team.
